### DISCUSSION TITLE
Custom function in datasource plugin doesn't show in data provider

### DISCUSSION THREAD

**Discussion by gmcnicol** [Budibase/budibase#8398] [Help]
Hello,

I've gone through the process of creating a custom data source plug-in. It has two custom functions that are functioning correctly.

When I add the custom datasource to my project and create a query using one of my custom functions called `fetchSomething`it returns preview data as expected.

I then try to configure it in a data provider component, and it's not in the list.

If I create a query from the same plugin but use the `Read`method, it gets listed in the data provider as an option.

Have I done something wrong?
Is this a bug?

**Comment by melohagan** [comment]
Hey @gmcnicol 

Thanks for having a go at building a custom datasource! 

Would you be able to provide some screenshots, and also a link to your repo is the plugin is public please?

**Comment by gmcnicol** [comment]
Hello, I've re-created the most basic of plugins: 

https://github.com/gmcnicol/custom-method-test

Here are some screenshots of my local admin screens:

First up is the plugin list in BB's main portal
![Screenshot from 2022-10-30 17-17-07](https://user-images.githubusercontent.com/1968845/198893069-5e4287be-0e34-44e2-b390-344d9bdc97fc.png)

Next, I've added the plugin as a data source, and configured two queries, one for read and one for readAdditional. I've tested the query, and it's displaying data, as you can see. 
![Screenshot from 2022-10-30 17-18-55](https://user-images.githubusercontent.com/1968845/198893078-5a77f7ea-1628-4121-8c41-07f776571d61.png)

![Screenshot from 2022-10-30 17-20-51](https://user-images.githubusercontent.com/1968845/198893084-f88e1f8d-1800-4222-b8d3-b2757ee5c5a4.png)

Adding a new data provider, trying to select ReadAdditional... it's not there.
![Screenshot from 2022-10-30 17-21-20](https://user-images.githubusercontent.com/1968845/198893088-81924f2b-b1dd-47eb-9bba-4924b89c3def.png)

Why?

Additional notes:
The documentation is inconsistent on how to initialise and watch the plug-in directories. 
`budi plugins --dev` does not seem to use the correct version of BB, as there are no plugins on the main portal. That's mentioned on one page. On the next page it's a different budi command, which seems to work, but it's all a bit... janky.

**Reply by gmcnicol** [reply]
@melohagan Additionally, I've tried on Firefox and Chrome. macOS and Ubuntu.

**Comment by melohagan** [comment]
Hey @gmcnicol 

I'm not quite sure what happened, but I was able to reproduce the bug you highlighted. Thanks for the screenshots!

I was able to fix it by deleting the **dist** folder and re-building the plugin.

I would suggest that you update the version number in the package json to `1.0.1` and push those changes up to Github. This will trigger another build via the GitHub workflow and the new version should hopefully be fixed!

**Reply by gmcnicol** [reply]
I am still seeing the problem. My plugins are added via file upload; internally, we won't be using GitHub at all. The plugins will be closed source, I'm afraid.

**Reply by melohagan** [reply]
This is quite strange. The only other thing I can think of, which seems unlikely to help, is to tidy the spacing:

![Screenshot 2022-11-01 at 08 28 12](https://user-images.githubusercontent.com/101575380/199191105-9c5f9c57-6afc-43c6-9525-5aca646678a8.png)

It's the only method with a leading space in 'query' and doesn't have the spacing around the curly braces.

**Reply by melohagan** [reply]
Or even remove the 'readAdditional' method from both the `datasource.ts` and `schema.json`. Build, verify that it works as expected.
Then re-add the 'readAdditional' method to the schema and datasource and maybe that'll fix it.

**Reply by gmcnicol** [reply]
If it were a syntax error,  then the previews would not work.

**Reply by gmcnicol** [reply]
I wonder if I could take a look at this in the code base. Where would one begin to look for where the system uploads plugins and where it lists the active data sources?

**Comment by melohagan** [comment]
Hey @gmcnicol 

I just found out today what might be causing the issue. Could you please try adding the **readable** property to your `ReadAdditional` function.
Similar to this: https://github.com/Budibase/budibase/pull/8593/files

**Reply by gmcnicol** [reply]
@melohagan sadly not. 
```
$ yarn build
yarn run v1.22.19
$ rollup -c

src/index.ts → dist/plugin.min.js...
[!] (plugin at position 2) ValidationError: "schema.query.readAdditional.readable" is not allowed
ValidationError: "schema.query.readAdditional.readable" is not allowed

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Reply by gmcnicol** [reply]
Adding that flag in the schema wouldn't have made much sense. The fact it works from a GitHub action and not a file upload suggests the problem is elsewhere. 

The pull request you link to is for a mainline integration, not a custom plugin, so I'm guessing the schema is slightly different. Looking at the rest of the file, there's a steps field etc., which isn't part of the plugin schema. The documentation also does not refer to this `readable` field. 

We've parked using Budibase in the project at work, but I'd still be keen to use it for my projects.

**Reply by melohagan** [reply]
Hey @gmcnicol I've pushed the readable schema change to version **2.1.24**
Could you update to the latest and see if that works for you now with the *readable* property set to true please?

**Comment by Hedronmx** [comment]
A quick workaround is to save the query  with the read method, go to the data provider section on your component, choose the now available method and then go back to your query and change the method to the one you intend to use.

edit- forgot to add that readable flag on schema file fixes the issue on 2.1.24

**Reply by melohagan** [reply]
Hey @Hedronmx This should be fixed now I believe. Add the readable property to your method, e.g. 

<img width="304" alt="Screenshot 2022-11-24 at 12 08 57" src="https://user-images.githubusercontent.com/101575380/203781038-e33b68fc-4aec-49f4-8b44-e58957b5c504.png">

**Reply by Hedronmx** [reply]
For some reason the readable tag works on one of my plugins but not in another. They are basically the same just using different apis.

I get the following error 

```
src/index.ts → dist/plugin.min.js...
[!] (plugin at position 2) ValidationError: "schema.query.getDevices.readable" is not allowed
ValidationError: "schema.query.getDevices.readable" is not allowed

error Command failed with exit code 1.

```

Schema looks like this

```
"getDevices": {
                "type": "fields",
                "readable": true,
                "fields": {
                    "deviceId": {
                        "type": "string",
                        "required": false
                    },
                    "resultsLimit": {
                        "type": "number",
                        "required": false
                    },
                    "database": {
                        "type": "string",
                        "required": true
                    }
                }
```

**Reply by melohagan** [reply]
Maybe it needs a clean build and re-import? Have you tried using the two plugins in the same app?

**Reply by Hedronmx** [reply]
Its failing on yarn build. Thats the output.

And yeah I am using both datasources on productions, using the above mentioned workaround.

**Reply by melohagan** [reply]
Hmm. Try deleting the dist folder and re-building the plugin.

Otherwise maybe make sure the Budibase CLI is up to date (npm install again). Init a new plugin and paste the schema/code into the new plugin.

**Reply by Hedronmx** [reply]
Will have to create a new plugin, deleting dist did not work. Thanks!

**Reply by gmcnicol** [reply]
This does not explain why it worked with GitHub actions and not file uploads. 

Is there code that reads the `readable` value in the designer? 

Why does the designer need it while the query editor doesn't?

What is the commit id for this fix?

Has the documentation been updated to have readable called out specifically?

I just don't think this is the solution. Allowing a new field in the JSON schema and then magic happens - software just isn't like that.

**Reply by melohagan** [reply]
The designer needs to to determine if the query provides data (e.g. a Read) and thus should be listed among the Data provider datasources, or if it does not provide data, e.g. a Create
